### PR TITLE
[trainer,vllm] feat: add a recipe to support reorder rollout

### PR DESCRIPTION
**Pull request by echo-rain**
### What does this PR do?
The long-tail effect in the current rollout phase is more pronounced in scenarios with long max response lengths. Therefore, we need a solution to address the long-tail effect in this scenario, thereby reducing the overall training time.

The PR strategy for this project comes from [RollPacker](https://arxiv.org/abs/2509.21009)

Add multiple requests during the data preparation phase. Once the required number of requests for training is met, end all training rollout requests and save them to a list. In a later step, re-add them in a long-tail batch.

experimental data in GRPO-Qwen3-8b_bf16-tp2-ep1 gen_batch_size=18 train_batch_size=16
normal mean ablation experimental group
image

### Checklist Before Starting

- [x] Search for similar PRs. Paste at least one query link here: [#2200 PR](https://github.com/volcengine/verl/pull/2200) [#2929 PR](https://github.com/volcengine/verl/pull/2929)
- [x] Format the PR title as `[{modules}] {type}: {description}` (This will be checked by the CI)
  - `{modules}` include `fsdp`, `megatron`, `sglang`, `vllm`, `rollout`, `trainer`, `ci`, `training_utils`, `recipe`, `hardware`, `deployment`, `ray`, `worker`, `single_controller`, `misc`, `perf`, `model`, `algo`, `env`, `tool`, `ckpt`, `doc`, `data`
  - If this PR involves multiple modules, separate them with `,` like `[megatron, fsdp, doc]`
  - `{type}` is in `feat`, `fix`, `refactor`, `chore`, `test`
  - If this PR breaks any API (CLI arguments, config, function signature, etc.), add `[BREAKING]` to the beginning of the title.
  - Example: `[BREAKING][fsdp, megatron] feat: dynamic batching`

### Test

example: examples/grpo_trainer/run_qwen3-8b_reorder.sh

### API and Usage Example

You can enable this feature by adding a batch size larger than the training batch size.
`data.train_batch_size=8
+data.gen_batch_size=10`

#### 1. Data preparation stage
Modify the existing data preparation logic, adding a new variable to configure the number of additional data records that need to be prepared. Add additional logic during data preparation to use identified long-tail requests to form batches.

#### 2. rollout post-process
The original `asycio.gather` check logic for the end of a rollout needs to be replaced with a check of the number of completed rollout requests. At the same time, unprocessed requests need to be marked as long-tail requests, added to a new long-tail request list, and the `abort` interface used to abort the request.

#### 3. reload req
Here we offer two options as solutions: adding in small amounts and multiple times, and adding all at once in multiple batches.

1. The meaning of adding a small number of requests is that long-tail requests from each step will be added to the next round of rollout<sup>1</sup>. After the requests are added, the long-tail request list is cleared<sup>2</sup>. However, due to the limited learning rate, requests that exhibited a long-tail effect in the previous round are likely to exhibit a long-tail effect in this round as well. This may lead to a rollback phenomenon, i.e., a rollback to the original implementation that uses the newly added requests. However, if the one-step-off-policy strategy of using the streaming return results of the previous round as the prompt input<sup>3</sup> can be used, the rollback phenomenon can be largely avoided.

2. Adding multiple requests together means that all long-tail requests are added to the training process to execute a complete training step, provided they form a complete batch. The drawback of this approach is that long-sequence requests that should be distributed across multiple training steps are concentrated into a single step, which may limit the learning rate and prevent effective learning of long-sequence representation capabilities.
normal step:

unfinished req step:

### Checklist Before Submitting

> [!IMPORTANT]
> Please check all the following items before requesting a review, otherwise the reviewer might deprioritize this PR for review.

- [x] Read the [Contribute Guide](https://github.com/volcengine/verl/blob/main/CONTRIBUTING.md).
- [x] Apply [pre-commit checks](https://github.com/volcengine/verl/blob/main/CONTRIBUTING.md#code-linting-and-formatting): `pre-commit install && pre-commit run --all-files --show-diff-on-failure --color=always`
- [x] Add / Update [the documentation](https://github.com/volcengine/verl/tree/main/docs).
- [x] Add unit or end-to-end test(s) to [the CI workflow](https://github.com/volcengine/verl/tree/main/.github/workflows) to cover all the code. If not feasible, explain why: ...
- [x] Once your PR is ready for CI, send a message in [the `ci-request` channel](https://verl-project.slack.com/archives/C091TCESWB1) in [the `verl` Slack workspace](https://join.slack.com/t/verl-project/shared_invite/zt-3855yhg8g-CTkqXu~hKojPCmo7k_yXTQ). (If not accessible, please try [the Feishu group (飞书群)](https://applink.larkoffice.com/client/chat/chatter/add_by_link?link_token=772jd4f1-cd91-441e-a820-498c6614126a).)

### CODE REVIEW

File: verl/utils/queue_util.py [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+
+import asyncio
+from asyncio import Queue
+from typing import Any
+
+import ray
+
+
+@ray.remote
+@ray.remote
```
[the remark this answers is not in the document]
Comment (echo-rain) [reply verl/utils/queue_util.py]: fixed

**Comment by yang-ybb** [comment]
@echo-rain when train_batch_size and gen_batch_size is small (like 8 and 10), rollpacker logic works. But when increase train_batch_size to 128, gen_batch_size to 130 and rollout.n to 16, vllm rollout time become very slow, seems related to these code
```
await self.queue.put.remote(
(
        raw_item,
        loop_output,
)
```

**Comment by echo-rain** [comment]
@yang-ybb It seems unlikely to be a queue issue. Although the queue lacks multi-process concurrency optimizations, this design is intended to prevent a large number of requests from completing within the same time window due to variable request completion times. 
With extremely large batches, the more probable problem is the slow inference speed of the rollout itself. 
However, the absence of parallel write logic does indeed limit queue write speed under high concurrency. Could you provide more relevant data so I can assess whether adding concurrency logic is necessary?

**Comment by yang-ybb** [comment]
@echo-rain ok~
1. use rollpacker, train_batch_size=128, gen_batch_size=129, rollout.n=16
first add time logs for `self.queue.put`, and results shows below. The execution time for the first few items is relatively normal, but the later ones take significantly longer, reaching 26 seconds. Total vllm rollout finish use 245s
```
Training Progress:   0%|          | 0/41500 [00:00<?, ?it/s]
(AgentLoopWorker pid=1704607) The image processor of type `Qwen2VLImageProcessor` is now loaded as a fast processor by default, even if the model checkpoint was saved with a slow processor. This is a breaking change and may produce slightly different outputs. To continue using the slow processor, instantiate this class with `use_fast=False`. Note that this behavior will be extended to all models in a future release. [repeated 7x across cluster]
(TaskRunner pid=1695273) ---------------global_steps 1 starts at 1766473912.1881988----------------
(AsyncvLLMServer pid=1699817) (EngineCore_0 pid=1701800) WARNING 12-23 07:11:30 [profiling.py:280] The sequence length (1344) is smaller than the pre-defined worst-case total number of multimodal tokens (32768). This may cause certain multi-modal inputs to fail during inference. To avoid this, you should increase `max_model_len` or reduce `mm_counts`. [repeated 7x across cluster]
(AsyncvLLMServer pid=1699817) WARNING 12-23 07:11:31 [__init__.py:1625] Default sampling parameters have been overridden by the model's Hugging Face generation config recommended from the model creator. If this is not intended, please relaunch vLLM instance with `--generation-config vllm`. [repeated 7x across cluster]
(RewardManagerWorker pid=1706240) The image processor of type `Qwen2VLImageProcessor` is now loaded as a fast processor by default, even if the model checkpoint was saved with a slow processor. This is a breaking change and may produce slightly different outputs. To continue using the slow processor, instantiate this class with `use_fast=False`. Note that this behavior will be extended to all models in a future release. [repeated 6x across cluster]
(AgentLoopWorker pid=1704612) You're using a Qwen2TokenizerFast tokenizer. Please note that with a fast tokenizer, using the `__call__` method is faster than using a method to encode the text followed by a call to the `pad` method to get a padded encoding.
(RewardManagerWorker pid=1706238) The image processor of type `Qwen2VLImageProcessor` is now loaded as a fast processor by default, even if the model checkpoint was saved with a slow processor. This is a breaking change and may produce slightly different outputs. To continue using the slow processor, instantiate this class with `use_fast=False`. Note that this behavior will be extended to all models in a future release. [repeated 2x across cluster]
(AgentLoopWorker pid=1704612) ============in agent_loop queue put cost 0.18525314331054688
(AgentLoopWorker pid=1704612) ============in agent_loop queue put cost 0.14816617965698242
(AgentLoopWorker pid=1704612) ============in agent_loop queue put cost 0.15647530555725098
(AgentLoopWorker pid=1704612) ============in agent_loop queue put cost 0.22381877899169922
(AgentLoopWorker pid=1704612) ============in agent_loop queue put cost 0.1943674087524414
(AgentLoopWorker pid=1704612) ============in agent_loop queue put cost 0.17458271980285645
(AgentLoopWorker pid=1704612) ============in agent_loop queue put cost 0.126600980758667
(AgentLoopWorker pid=1704612) ============in agent_loop queue put cost 0.1065363883972168
(AgentLoopWorker pid=1704612) ============in agent_loop queue put cost 0.11809802055358887
(AgentLoopWorker pid=1704612) ============in agent_loop queue put cost 0.10920262336730957
(AgentLoopWorker pid=1704612) ============in agent_loop queue put cost 0.08970022201538086
(AgentLoopWorker pid=1704612) ============in agent_loop queue put cost 0.08512425422668457
(AgentLoopWorker pid=1704612) ============in agent_loop queue put cost 0.11975407600402832
(AgentLoopWorker pid=1704612) ============in agent_loop queue put cost 0.0790250301361084
(AgentLoopWorker pid=1704612) ============in agent_loop queue put cost 0.10097265243530273
(AgentLoopWorker pid=1704612) ============in agent_loop queue put cost 0.06738114356994629
(AgentLoopWorker pid=1704607) ============in agent_loop queue put cost 0.08369851112365723 [repeated 58x across cluster]
(AgentLoopWorker pid=1704604) You're using a Qwen2TokenizerFast tokenizer. Please note that with a fast tokenizer, using the `__call__` method is faster than using a method to encode the text followed by a call to the `pad` method to get a padded encoding. [repeated 6x across cluster]
(AgentLoopWorker pid=1704605) ============in agent_loop queue put cost 0.30245041847229004 [repeated 152x across cluster]
(AgentLoopWorker pid=1704605) ============in agent_loop queue put cost 3.4407339096069336 [repeated 148x across cluster]
(AgentLoopWorker pid=1704605) ============in agent_loop queue put cost 7.087812662124634 [repeated 199x across cluster]
(AgentLoopWorker pid=1704605) ============in agent_loop queue put cost 10.952093839645386 [repeated 260x across cluster]
(AgentLoopWorker pid=1704610) ============in agent_loop queue put cost 14.84812307357788 [repeated 257x across cluster]
(AgentLoopWorker pid=1704610) ============in agent_loop queue put cost 18.69814658164978 [repeated 237x across cluster]
(AgentLoopWorker pid=1704610) ============in agent_loop queue put cost 22.471090078353882 [repeated 200x across cluster]
(AgentLoopWorker pid=1704606) ============in agent_loop queue put cost 26.274360418319702 [repeated 250x across cluster]
(AgentLoopWorker pid=1704604) ============in agent_loop queue put cost 26.876274585723877 [repeated 232x across cluster]
(TaskRunner pid=1695273) ---_run_reorder_worker_generate_sequences tasks finish: 1766473980.4850829. cost 65.91393113136292
(AgentLoopWorker pid=1704607) ============in agent_loop queue put cost 26.489635467529297 [repeated 39x across cluster]
(TaskRunner pid=1695273) ---_run_reorder_worker_generate_sequences queue clear finish: 1766474156.4789267. cost 241.90777111053467
(TaskRunner pid=1695273) ---------------global_steps 1 generate_sequences finished at 1766474157.9028032 costs 245.71460437774658----------------
(TaskRunner pid=1695273) ---------------global_steps 1 rewards finished at 1766474160.5839174 costs 248.39571857452393----------------
(TaskRunner pid=1695273) on_policy=True, skip compute_old_log_prob.
(TaskRunner pid=1695273) ---------------global_steps 1 old_log_prob finished at 1766474160.584271 costs 248.39607214927673----------------
(TaskRunner pid=1695273) ---------------global_steps 1 cal_ref finished at 1766474160.5842805 costs 248.3960816860199----------------
(TaskRunner pid=1695273) ---------------global_steps 1 cal_adv finished at 1766474169.90291 costs 257.71471118927----------------
(TaskRunner pid=1695273) ---------------global_steps 1 cal_critic finished at 1766474169.9029684 costs 257.7147696018219----------------
```
second shows gpu utl
image

2. no rollpacker, train_batch_size=128, rollout.n=16
Total vllm rollout finish use 35s
```
Training Progress:   0%|          | 0/41500 [00:00<?, ?it/s]
(TaskRunner pid=1715642) ---------------global_steps 1 starts at 1766474490.5935266----------------
(AgentLoopWorker pid=1724728) The image processor of type `Qwen2VLImageProcessor` is now loaded as a fast processor by default, even if the model checkpoint was saved with a slow processor. This is a breaking change and may produce slightly different outputs. To continue using the slow processor, instantiate this class with `use_fast=False`. Note that this behavior will be extended to all models in a future release.
(RewardManagerWorker pid=1726250) The image processor of type `Qwen2VLImageProcessor` is now loaded as a fast processor by default, even if the model checkpoint was saved with a slow processor. This is a breaking change and may produce slightly different outputs. To continue using the slow processor, instantiate this class with `use_fast=False`. Note that this behavior will be extended to all models in a future release. [repeated 8x across cluster]
(AgentLoopWorker pid=1724730) You're using a Qwen2TokenizerFast tokenizer. Please note that with a fast tokenizer, using the `__call__` method is faster than using a method to encode the text followed by a call to the `pad` method to get a padded encoding.
(RewardManagerWorker pid=1727550) The image processor of type `Qwen2VLImageProcessor` is now loaded as a fast processor by default, even if the model checkpoint was saved with a slow processor. This is a breaking change and may produce slightly different outputs. To continue using the slow processor, instantiate this class with `use_fast=False`. Note that this behavior will be extended to all models in a future release. [repeated 7x across cluster]
(AgentLoopWorker pid=1724733) You're using a Qwen2TokenizerFast tokenizer. Please note that with a fast tokenizer, using the `__call__` method is faster than using a method to encode the text followed by a call to the `pad` method to get a padded encoding. [repeated 6x across cluster]
(TaskRunner pid=1715642) ---------------global_steps 1 generate_sequences finished at 1766474526.0778503 costs 35.48432374000549----------------
(AsyncvLLMServer pid=1719964) (EngineCore_0 pid=1721927) WARNING 12-23 07:21:26 [profiling.py:280] The sequence length (1344) is smaller than the pre-defined worst-case total number of multimodal tokens (32768). This may cause certain multi-modal inputs to fail during inference. To avoid this, you should increase `max_model_len` or reduce `mm_counts`. [repeated 7x across cluster]
(AsyncvLLMServer pid=1719964) WARNING 12-23 07:21:27 [__init__.py:1625] Default sampling parameters have been overridden by the model's Hugging Face generation config recommended from the model creator. If this is not intended, please relaunch vLLM instance with `--generation-config vllm`. [repeated 7x across cluster]
(TaskRunner pid=1715642) ---------------global_steps 1 rewards finished at 1766474528.536392 costs 37.9428653717041----------------
(TaskRunner pid=1715642) on_policy=True, skip compute_old_log_prob.
(TaskRunner pid=1715642) ---------------global_steps 1 old_log_prob finished at 1766474528.5366735 costs 37.94314694404602----------------
(TaskRunner pid=1715642) ---------------global_steps 1 cal_ref finished at 1766474528.5366833 costs 37.943156719207764----------------
(TaskRunner pid=1715642) ---------------global_steps 1 cal_adv finished at 1766474537.8688495 costs 47.275322914123535----------------
(TaskRunner pid=1715642) ---------------global_steps 1 cal_critic finished at 1766474537.8689034 costs 47.27537679672241----------------
```
image

**Comment by echo-rain** [comment]
@yang-ybb got it thank you.
Based on the data, the queue's insertion time is not as expected. I will try to reproduce this phenomenon locally and attempt to fix it.

**Comment by echo-rain** [comment]
@yang-ybb To be precise, this performance issue isn't caused by concurrency. The root cause is that the `put` method of the queue serializes the data, and the large size of the batch data class causes this serialization to be very slow. 
Passing pointers in the queue might be a potentially better solution.
I will fix this problem as soon as possible.

**Comment by yang-ybb** [comment]
@echo-rain yes, it may be put large data use ray actor. time cost normal after change to self.queue.put.remote(1), and put results in each AgentLoopWorker. and in AgentLoopManager, use worker.get_results.remote() can gather all results to main thread.
```
# AgentLoopWorker._run_agent_loop
if self.reorder_rollout:
    await self.queue.put.remote(1)
    self.finish_results.append((raw_item, loop_output))
    return
else:
    return loop_output
```

```
# AgentLoopManager._run_reorder_worker_generate_sequences

- hybird_out = ray.get(self.queue.clear.remote())
+ hybird_out = ray.get([
    worker.get_results.remote()
    for worker in self.agent_loop_workers
])
```

**Comment by echo-rain** [comment]
@yang-ybb Great idea, thanks for the suggestion. However, in practice, any remote call will serialize the transmitted data. Perhaps a better implementation would be to store the larger tensor in the main process and mark it with some kind of pointer, then queue the pointer for transmission.

**Comment by cboss6** [comment]
@echo-rain
I tried reorder_rollout with Qwen3-8B and found reorder_rollout affects the convergence of the training：
Clipboard_Screenshot_1767701854
Clipboard_Screenshot_1767701879

The normal behavior should be(the only difference in script is whether gen_train_size equals to train_batch_size):
Clipboard_Screenshot_1767702062
Clipboard_Screenshot_1767702078

Do you have any comments? As far as I know, theoretically RollPacker shouldn’t affect the convergence of RL training.

**Comment by echo-rain** [comment]
@cboss6 Thank you for your comment. In fact, we have already noticed this issue, which is why this PR has not been merged for so long.The results show that RollPacker does have an impact on training performance.In my opinion, adding long requests to the same batch leads to two potential problems. First, the aggregation of long requests causes a sharp increase in length-related penalties. Second, long requests are more easily truncated, resulting in generally lower reward scores.


Of course, this is just my conjecture. If you have better conclusions or solutions, please feel free to discuss them with me.
